File: frontend/de/unika/ipd/grgen/ir/stmt/invocation/ExternalProcedureMethodInvocation.java
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 6.5
 * Copyright (C) 2003-2022 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3 (see LICENSE.txt included in the packaging of this file)
 * www.grgen.net
 */

/**
 * @author Edgar Jakumeit
 */

package de.unika.ipd.grgen.ir.stmt.invocation;

import de.unika.ipd.grgen.ir.NeededEntities;
import de.unika.ipd.grgen.ir.executable.ExternalProcedure;
import de.unika.ipd.grgen.ir.executable.ProcedureBase;
import de.unika.ipd.grgen.ir.expr.Expression;
import de.unika.ipd.grgen.ir.expr.Qualification;
import de.unika.ipd.grgen.ir.pattern.GraphEntity;
import de.unika.ipd.grgen.ir.pattern.Variable;

/**
 * An external procedure method invocation.
 */
public class ExternalProcedureMethodInvocation extends ProcedureInvocationBase
{
	/** The owner of the procedure method. */
	private Qualification ownerQual;
	private Variable ownerVar;

	/** The procedure of the procedure method invocation expression. */
	protected ExternalProcedure externalProcedure;

	public ExternalProcedureMethodInvocation(Qualification ownerQual, ExternalProcedure externalProcedure)
	{
		super("external procedure method invocation");

		this.ownerQual = ownerQual;
		this.externalProcedure = externalProcedure;
	}

	public ExternalProcedureMethodInvocation(Variable ownerVar, ExternalProcedure externalProcedure)
	{
		super("external procedure method invocation");

		this.ownerVar = ownerVar;
		this.externalProcedure = externalProcedure;
	}

	public Qualification getOwnerQual()
	{
		return ownerQual;
	}

	public Variable getOwnerVar()
	{
		return ownerVar;
	}

	@Override
	public ProcedureBase getProcedureBase()
	{
		return externalProcedure;
	}

	public ExternalProcedure getExternalProc()
	{
		return externalProcedure;
	}

	/** @see de.unika.ipd.grgen.ir.expr.Expression#collectNeededEntities() */
	@Override
	public void collectNeededEntities(NeededEntities needs)
	{
		if(ownerQual != null) {
			ownerQual.collectNeededEntities(needs);
			if(ownerQual.getOwner() != null) {
				if(ownerQual.getOwner() instanceof GraphEntity)
					needs.add((GraphEntity)ownerQual.getOwner());
			}
		} else {
			if(!isGlobalVariable(ownerVar))
				needs.add(ownerVar);
		}
		for(Expression child : getWalkableChildren()) {
			child.collectNeededEntities(needs);
		}
	}
}

File: frontend/de/unika/ipd/grgen/ir/stmt/invocation/ExternalProcedureMethodInvocation.cs
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 8.1
 * Copyright (C) 2003-2026 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3, some components/parts use different licenses (see LICENSE.txt included in the packaging of this file)
 * www.grgen.de / www.grgen.net
 */

/// <summary>
/// @author Edgar Jakumeit
/// </summary>

namespace de.unika.ipd.grgen.ir.stmt.invocation
{
using NeededEntities = de.unika.ipd.grgen.ir.NeededEntities;
using ExternalProcedure = de.unika.ipd.grgen.ir.executable.ExternalProcedure;
using ProcedureBase = de.unika.ipd.grgen.ir.executable.ProcedureBase;
using Expression = de.unika.ipd.grgen.ir.expr.Expression;
using Qualification = de.unika.ipd.grgen.ir.expr.Qualification;
using GraphEntity = de.unika.ipd.grgen.ir.pattern.GraphEntity;
using Variable = de.unika.ipd.grgen.ir.pattern.Variable;

/// <summary>
/// An external procedure method invocation.
/// </summary>
public class ExternalProcedureMethodInvocation : ProcedureInvocationBase
{
	/// <summary>
	/// The owner of the procedure method. </summary>
	private Qualification ownerQual;
	private Variable ownerVar;

	/// <summary>
	/// The procedure of the procedure method invocation expression. </summary>
	protected internal ExternalProcedure externalProcedure;

	public ExternalProcedureMethodInvocation(Qualification ownerQual, ExternalProcedure externalProcedure)
		: base("external procedure method invocation")
	{

		this.ownerQual = ownerQual;
		this.externalProcedure = externalProcedure;
	}

	public ExternalProcedureMethodInvocation(Variable ownerVar, ExternalProcedure externalProcedure)
		: base("external procedure method invocation")
	{

		this.ownerVar = ownerVar;
		this.externalProcedure = externalProcedure;
	}

	public virtual Qualification OwnerQual
	{
		get
		{
			return ownerQual;
		}
	}

	public virtual Variable OwnerVar
	{
		get
		{
			return ownerVar;
		}
	}

	public override ProcedureBase ProcedureBase
	{
		get
		{
			return externalProcedure;
		}
	}

	public virtual ExternalProcedure ExternalProc
	{
		get
		{
			return externalProcedure;
		}
	}

	/// <seealso cref="de.unika.ipd.grgen.ir.expr.Expression.collectNeededEntities() "/>
	public override void CollectNeededEntities(NeededEntities needs)
	{
		if(ownerQual != null)
		{
			ownerQual.CollectNeededEntities(needs);
			if(ownerQual.Owner != null)
			{
				if(ownerQual.Owner is GraphEntity)
					needs.Add((GraphEntity)ownerQual.Owner);
			}
		}
		else
		{
			if(!IsGlobalVariable(ownerVar))
				needs.Add(ownerVar);
		}
		foreach(Expression child in WalkableChildren)
			child.CollectNeededEntities(needs);
	}
}

}
